movement/analysis/kinematics.py

```diff
@@ -161,6 +163,94 @@ def _compute_approximate_time_derivative(
+    forwards (in the rostral direction). As the forward direction may
+    differ between coordinate systems, the front keypoint is used ...,
+    when present. Otherwise, we assume that coordinates are given in the
+    image coordinate system (where the origin is located in the top-left).
+
+    Parameters
+    ----------
+    data : xarray.DataArray
+        The input data representing position. This must contain
+        the two chosen keypoints corresponding to the left and
+        right of the head.
+    left_keypoint : str
+        Name of the left keypoint, e.g., "left_ear"
+    right_keypoint : str
+        Name of the right keypoint, e.g., "right_ear"
+    front_keypoint : str | None
+        (Optional) Name of the front keypoint, e.g., "nose".
+
+    Returns
+    -------
+    xarray.DataArray
+        An xarray DataArray representing the head direction vector,
+        with dimensions matching the input data array, but without the
+        ``keypoints`` dimension.
+
+    """
+    # Validate input dataset
+    _validate_type_data_array(data)
+    _validate_time_keypoints_space_dimensions(data)
+
+    if left_keypoint == right_keypoint:
+        raise log_error(
+            ValueError, "The left and right keypoints may not be identical."
+        )
+    if len(data.space) != 2:
+        raise log_error(
+            ValueError,
+            "Input data must have 2 (and only 2) spatial dimensions, but "
+            f"currently has {len(data.space)}.",
+        )
+
+    # Select the right and left keypoints
+    head_left = data.sel(keypoints=left_keypoint, drop=True)
+    head_right = data.sel(keypoints=right_keypoint, drop=True)
+
+    # Initialize a vector from right to left ear, and another vector
+    # perpendicular to the X-Y plane
+    right_to_left_vector = head_left - head_right
+    perpendicular_vector = np.array([0, 0, -1])
+
+    # Compute cross product
+    head_vector = head_right.copy()
+    head_vector.values = np.cross(right_to_left_vector, perpendicular_vector)[
+        :, :, :-1
+    ]
+
+    # Check computed head_vector is pointing in the same direction as vector
+    # from head midpoint to snout
+    if front_keypoint:
+        head_front = data.sel(keypoints=front_keypoint, drop=True)
+        head_midpoint = (head_right + head_left) / 2
+        mid_to_front_vector = head_front - head_midpoint
+        dot_product_array = (
+            convert_to_unit(head_vector.sel(individuals=data.individuals[0]))
+            * convert_to_unit(mid_to_front_vector).sel(
+                individuals=data.individuals[0]
+            )
+        ).sum(dim="space")
+        median_dot_product = float(dot_product_array.median(dim="time").values)
+        if median_dot_product < 0:
+            perpendicular_vector = np.array([0, 0, 1])
+            head_vector.values = np.cross(
+                right_to_left_vector, perpendicular_vector
+            )[:, :, :-1]
+
+    return head_vector
```

Comment:
```suggestion
    # Validate input dataset
    _validate_type_data_array(data)
    _validate_time_keypoints_space_dimensions(data)
    if len(data.space) != 2:
        raise log_error(
            ValueError,
            "Input data must have 2 (and only 2) spatial dimensions, but "
            f"currently has {len(data.space)}.",
        )

    # Validate input keypoints
    if left_keypoint == right_keypoint:
        raise log_error(
            ValueError, "The left and right keypoints may not be identical."
        )

    # Define right-to-left vector (as a data array)
    right_to_left_vector = data.position.sel(
        keypoints=left_keypoint, drop=True
    ) - data.position.sel(keypoints=right_keypoint, drop=True)

    # Define upward vector (as a data array)
    # default: negative z direction in the image coordinate system
    upward_vector = xr.DataArray(
        np.array(upward_vector).reshape(1, -1), dims=["time", "space"]
    )

    # Compute forward direction as the cross product
    # (right-to-left) cross (forward) = up
    forward_vector = xr.cross(
        right_to_left_vector, upward_vector, dim="space"
    )[:, :, :-1]  # keep only the first 2 dimensions of the result

    # Return unit vector
    return forward_vector / compute_norm(forward_vector)
```

Reply:
Some bits to note:
- I made `right_to_left_vector` a data array rather than a dataset, because I think we can drop the difference in confidence array.
- We can use `xarray`'s `cross` product to keep the result as a data array (I think before it was just a numpy array?)
- I think it is more clear to return an unit vector for the head/forward vector, since the length of the vector doesn't really have a meaning (it will be the product of the norms of the two vectors in the product)